src/data/real/liouville.lean

```diff
@@ -28,34 +24,143 @@ A Liouville number is a real number `x` such that for every natural number `n`,
+  -- A few useful inequalities
   have qR0 : (0 : ℝ) < q := int.cast_pos.mpr (zero_lt_one.trans q1),
   have b0 : (b : ℝ) ≠ 0 := ne_of_gt (nat.cast_pos.mpr bN0),
   have bq0 : (0 : ℝ) < b * q := mul_pos (nat.cast_pos.mpr bN0) qR0,
+  -- At a1, clear denominators...
   replace a1 : abs (a * q - b * p) * q ^ (b + 1) < b * q, by
     rwa [div_sub_div _ _ b0 (ne_of_gt qR0), abs_div, div_lt_div_iff (abs_pos.mpr (ne_of_gt bq0))
-      (pow_pos qR0 _), abs_of_pos bq0, one_mul, ← int.cast_pow, ← int.cast_mul, ← int.cast_coe_nat,
-      ← int.cast_mul, ← int.cast_mul, ← int.cast_sub, ← int.cast_abs, ← int.cast_mul, int.cast_lt]
-        at a1,
+      (pow_pos qR0 _), abs_of_pos bq0, one_mul,
+      -- ... and revert to integers
+       ← int.cast_pow, ← int.cast_mul, ← int.cast_coe_nat, ← int.cast_mul, ← int.cast_mul,
+       ← int.cast_sub, ← int.cast_abs, ← int.cast_mul, int.cast_lt] at a1,
+  -- At a0, clear denominators...
   replace a0 : ¬a * q - ↑b * p = 0, by
     rwa [ne.def, div_eq_div_iff b0 (ne_of_gt qR0), mul_comm ↑p, ← sub_eq_zero_iff_eq,
+      -- ... and revert to integers
       ← int.cast_coe_nat, ← int.cast_mul, ← int.cast_mul, ← int.cast_sub, int.cast_eq_zero] at a0,
+  -- Actually, `q` is a natural number
   lift q to ℕ using (zero_lt_one.trans q1).le,
+  -- Looks innocuous, but we now have an integer with non-zero absolute value: this is at
+  --least one away from zero.  The gain here is what gets the proof going.
   have ap : 0 < abs (a * ↑q - ↑b * p) := abs_pos.mpr a0,
+  -- Actually, the absolute value of an integer is a natural number
   lift (abs (a * ↑q - ↑b * p)) to ℕ using (abs_nonneg (a * ↑q - ↑b * p)),
+  -- At a1, revert to natural numbers
   rw [← int.coe_nat_mul, ← int.coe_nat_pow, ← int.coe_nat_mul, int.coe_nat_lt] at a1,
+  -- Recall this is by contradiction: we obtained the inequality `b * q ≤ x * q ^ (b + 1)`, so
+  -- we are done.
   exact not_le.mpr a1 (nat.mul_lt_mul_pow_succ (int.coe_nat_pos.mp ap) (int.coe_nat_lt.mp q1)).le,
 end
 
-end is_liouville
+end liouville
+
+open polynomial metric set real ring_hom
+
+/-- This lemma collects the properties needed to prove `exists_pos_real_of_irrational_root`.
+It is stated in more general form than needed: in the intended application, `Z = ℤ`, `N = ℕ`,
+`R = ℝ`, `d a = a ^ f.nat_degree`, `j z a  = z / (a + 1)`, `f ∈ ℤ[x]`, `α` is an irrational
+root of `f`, `ε` is small, `M` is a bound on the Lipschitz constant of `f` near `α`, `n` is
+the degree of the polynomial `f`.
+-/
+lemma exists_one_le_pow_mul_dist {Z N R : Type*} [metric_space R]
+  {d : N → ℝ} {j : Z → N → R} {f : R → R} {α : R} {ε M : ℝ}
+-- denominators are positive
+  (d0 : ∀ (a : N), 1 ≤ d a)
+  (e0 : 0 < ε)
+--function is Lipschitz at α
+  (B : ∀ ⦃y : R⦄, y ∈ closed_ball α ε → dist (f α) (f y) ≤ (dist α y) * M)
+-- clear denominators
+  (L : ∀ ⦃z : Z⦄, ∀ ⦃a : N⦄, j z a ∈ closed_ball α ε → 1 ≤ (d a) * dist (f α) (f (j z a))) :
+  ∃ e : ℝ, 0 < e ∧ ∀ (z : Z), ∀ (a : N), 1 ≤ (d a) * (dist α (j z a) * e) :=
```

Comment:
I have a psychological issue with using `e` for a large number. What about `C`? Same thing in the next lemma, what is named `ε` there is probably rather large, so could you name it `C`? Another solution would be to present your inequality as `ε ≤ (d a) * dist α (j z a)`, which is probably easier to read, so I'd go for this one.

Reply:
I will go for `M` instead of `ε` and leave it on the same side of the inequality.  I am worried of using `C` for potential interactions with `polynomial.C`, the inclusion of a ring `R` into `polynomial R`.

Reply:
Actually, I went for `A` since `M` already appeared in the statement.